src/PowerAuth/v3/AuthenticationServiceV3.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2025 Wultra s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "AuthenticationServiceV3.h"
+#include "FunctionsV3.h"
+#include "../model/Constants.h"
+#include "../common/CommonFunctions.h"
+#include "../request/RequestBuilder.h"
+#include "../HttpHeaderHelper.h"
+
+namespace powerAuth {
+namespace v3 {
+
+#define LOCK_GUARD() std::lock_guard<std::recursive_mutex> _lock_guard(*_lock)
+
+AuthenticationServiceV3::AuthenticationServiceV3(const ContextPtr& context) :
+    Service("AuthenticationServiceV3", context->getSharedMutexPtr()),
+    _weak_context(context),
+    _configuration(context->getConfigurationPtr()),
+    _session_data(context->getSessionDataPtr()),
+    _key_provider(context->getKeyProviderPtr())
+{
+}
+
+IServicePtr AuthenticationServiceV3::asService()
+{
+    return shared_from_this();
+}
+
+HttpHeader AuthenticationServiceV3::calculateOnlineAuthenticationHeader(const Credentials& credentials,
+                                               const OnlineAuthenticationData& auth_data,
+                                               const cc7::ByteRange& body)
+{
+    LOCK_GUARD();
+    
+    if (_session_data->hasRegistrationData()) {
+        throw Exception(EC_WrongActivationState, "Authentication header calculation is not allowed during activation registration");
+    }
+    
+    if (_session_data->hasUpgradeData()) {
+        throw Exception(EC_WrongActivationState, "Authentication header calculation is not allowed during pending protocol upgrade");
+    }
```

Comment:
I think there will be exception in the upgrade process itself. The request that initiate the upgrade process is signed with V3 signature while `hasUpgradeData()` will be true at the time of signature computation. So, I recommend you to also test `auth_data.allowedInUpgrade` flag.